Newsletters: add #46 (2019-05-14)

harding opened the pull request:
A newsletter with pretensions of being a book.

moneyball commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+  is made to allow for batch verification of multiple signatures, which
+  [bip-schnorr][] estimates can speed up verification by FIXMEx over
+  checking each signature independently.
+
```
Batch validation speedup, for a block of 1k-10k Schnorr signatures, is shown to be 2.2-2.6x in the Schnorr BIP[1]. However, that data included pubkey decompression. If that is factored out, then a speedup of 3-4x is anticipated for the same # of transactions[2].

[1] https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki
[2] https://bitcoin.stackexchange.com/questions/85213/when-schnorr-signatures-are-part-of-bitcoin-will-it-be-possible-validate-each-bl/85215

moneyball commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+        (1) <A pubkey> OP_CHECKSIGVERIFY <B pubkey> OP_CHECKSIG
+        (2) <A pubkey> OP_CHECKSIGVERIFY <C pubkey> OP_CHECKSIG
+        (3) <B pubkey> OP_CHECKSIGVERIFY <C pubkey> OP_CHECKSIG
+
```
Do we want to point out that this will (unfortunately) be heavier weight and more costly than using traditional P2WSH OP_CHECKMULTISIG due to having to broadcast 1 or more MAST hashes?

harding replied:
I haven't done the exact math yet, but quick mental math told me the main example in the previous subsection (the HTLC) is going to come out with one of the branches more expensive than P2WSH, and that's why I added a note there, "this means that more
complex scripts than our simple HTLC can gain much larger efficiency
improvements than seen above (see an article focused on [MAST][] for
more information)."

I think being explicit there and showing how small trees can result in higher costs will give us a pass for avoiding details here.  The objective of this example is to show how k-of-n can become multiple k-of-k, and there are many cases for larger values of n where this will be more cost effective than full scripts (plus there's always key-path spending).

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+another time and focus here on how the two proposed BIPs can make
+existing uses of Bitcoin work even better than they do today.
+
+### What's not in the proposals
```
I do think it breaks the flow of the section a bit to put it up front---but it would also break the flow a bit no matter where we put it.  The advantage of the current placement is that is clears up some misconceptions before we get into details and it allows the section to end on a high note.

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+  desirable before we start a debate about the best way to activate
+  them.
+
+- **No changes to existing operations for wallets:** all Taproot
```
Do you think it shouldn't be?  It seems to fit here IMO, but I admit I did originally mention this in the single-sig subsection below.

moneyball replied:
I interpret the section "What's not in the proposal" to be features or changes that might have been anticipated but will be in future proposals or soft forks. This bullet point, which is communicating that no changes must be made to a wallet due to this soft fork, seems to better fits under a "services implementation details of this proposal."

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+a 0x02 or 0x03 to allow validators to reconstruct the key's Y-coordinate
+on the secp256k1 elliptic curve; in bip-taproot, the value of this byte
+is reduced by two so that 0x02 becomes 0x00 and 0x03 becomes 0x01,
+although the meaning is exactly the same.  Also the witness version is
```
I don't know for sure, which is why I didn't say.  My guess is that it allows the value to be dependent on a single bit (the low bit), allowing the other 7 bits to be used in subsequent soft forks.

sipa replied:
Yeah, just picking the first 2 values available.

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+However, the costs are distributed differently---it costs more to create
+a Taproot output than a P2WPKH output and less to spend it, which may
+help contribute towards keeping the size of the UTXO set manageable.
+
```
I don't think that's correct.  An algorithm that evaluates whether it makes more economic sense to create a change output or spend change to fees should be considering not just the cost to create the change but also the cost to later spend it.  As this paragraph explains, these costs are essentially the same for P2WPKH and P2TR because the higher output costs are mitigated by lower input costs.

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+vbytes); if a non-default is used, it is 65 bytes (16.25
+vbytes[^vbytes-decimal]).  Overall, this makes the cost to create and
+spend a Taproot single-sig output about 1% more expensive than P2WPKH.
+However, the costs are distributed differently---it costs more to create
```
> I have a slight concern that many wallets won't upgrade because "why do the work to pay 1% more fees."

Me too.  However there are huge advantages for multisig users and advanced script users, so they can lead the way.  More importantly, the reduced costs of multisig and advanced scripts will help attract users to more secure methods of using Bitcoin than single-sig, reducing the number of single-sig users.  When enough people are using Taproot anyway, then there's privacy advantage to using it for single-sig too, which may outweigh the 1% difference.

> if we look at common wallet spends such as 2 input, 1 output or 3 input, 2 output, those become meaningfully more economical using Taproot.

Right, but somebody had to pay the fees to create those inputs' prevouts and that probably affected the amount of money you received.  This is like insisting that companies sent you their bills with business return envelopes so you don't have to provide your own stamp; that may seem cheaper to you, but companies are just going to charge you the cost of postage anyway.

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+prove the transaction is an authorized spend of a UTXO.
+
+Most notably, the hash used goes from the legacy protocol's
+double-SHA256 (SHA256d) to a single SHA256 operation.  The extra hashing
```
I haven't measured, but some napkin math says that if all 412 million txes on the block chain have an average of 2 inputs each with 1 signature, this could save a 10 to 20 minutes off of IBD.  However, the transaction digest formula now includes hashes in other places, so I think the total number of hashes increases (although the amount of data being hashed decreases), which makes it beyond the capacity of my napkin to clearly evaluate.

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+consensus changes.  This is because the signature algorithm is only
+looking for a single pubkey and a single signature that are valid under
+rules described in the single-sig subsection above.  That said, of the
+methods known, the [MuSig protocol][] is probably the best studied
```
Right, I was hoping to avoid those details here because they're not directly relevant to the bip-taproot proposal.  The point I want to make is that you can use any protocol you want as long is it produces a valid sig for a valid pubkey.  There are ways known for creating aggregated pubkeys and signatures without interaction---they're just not secure against your counter signers.  Maybe that isn't a problem for some application, so they can use that protocol.  Or, as you say, maybe improvements will be made.  I don't think those specifics matter in the context of bip-taproot; what matters is that we know aggregation is possible and that it's up to individiual application developers to choose how they want to do it and what tradeoffs they want to make.

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+complex scripts than our simple HTLC can gain much larger efficiency
+improvements than seen above (see an article focused on [MAST][] for
+more information).
+
```
This is great info, but if we go with my previous comment about not getting into MuSig details, then I don't think this fits very well in this section.

I was thinking that, when we finish the 24 weeks of bech32, we might keep the long-term focus section and convert it to describing a detail about schnorr/taproot/tapscript each week.  I think this would work well in something like that.  Alternatively, if you don't want to wait three months, you could post a blog (I'd be happy to help, if you want).

moneyball replied:
Sounds good and either approach works for me.

kanzure commented:
Hi, there were some slightly technical talks at MCC over the weekend that might warrant inclusion http://diyhpl.us/wiki/transcripts/magicalcryptoconference/2019/

harding commented:
@kanzure thanks!  I don't have the time to include anything else non-trivial in this week's newsletter but I've got your link open in a tab and I plan to go through the transcripts as soon as I finish the edits for this week.  (As always, thanks for typing the transcripts; their existence makes writing summaries for the newsletter so much easier for me.)

jnewbery commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+
+Last week, Pieter Wuille [emailed][] the Bitcoin-Dev mailing list with
+links to two proposed BIPs.  The first, [bip-taproot][], permits
+spending using either a Schnorr-style signature or merklized script. The
```
Merklized is consistent with Dave's previous articles: https://gist.github.com/maaku/f7b2e710c53f601279549aa74eeb5368 (I also questioned this when Dave wrote that article but was convinced the the verb form of Merkle is merklize)

ajtowns commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,485 @@
+  previously-valid things invalid, this maximal validity
+  today will allow maximal flexibility in future soft forks.  Of course,
+  receivers get to choose what scripts they receive payments to, so no
+  one who likes their money will include an `OP_SUCCESSn` opcode in
```
It's OP_SUCCESSx in bip-tapscript as a placeholder for OP_SUCCESS80, OP_SUCCESS98, OP_SUCCESS126, etc.

harding replied:
Although I think it was technically correct, I ended up removing the `n` from this text because it was a bit confusing in the reduced context of a summary.

ajtowns commented on _posts/en/newsletters/2019-05-14-newsletter.md:
Putting anything in the annex is non-standard (so that it's easy to soft-fork in meaningful stuff later), so it's not _that_ easy to stuff data in there.

jnewbery commented on _posts/en/newsletters/2019-05-14-newsletter.md:
Suggest removing "to the database". "forward payments to the database" could be confusing.

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
Produces an emdash in this version of Markdown.

ajtowns replied:
yeah; having it inside the stars confused me, i think

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
Oh, duh.  :man_facepalming:

sipa commented on _posts/en/newsletters/2019-05-14-newsletter.md:
Generally you should only end up with "unpaired" nodes in the tree when you're already stuffing the tree with random values for privacy (for example to hide how many entries are in the tree).

Without those, you won't have unpaired nodes, as you'd instead have the lone node moved up one level.

jnewbery commented on _posts/en/newsletters/2019-05-14-newsletter.md:
I don't understand the "(the same operation used in BIP32 HD key derivation)" and would suggest you drop it.

jnewbery commented on _posts/en/newsletters/2019-05-14-newsletter.md:
I don't think this is necessary. You can just build an unbalanced tree.

sipa commented on _posts/en/newsletters/2019-05-14-newsletter.md:
Unless you're stuffing the tree intentionally with random values for privacy, you can't have "unpaired" nodes; you'd just have that node in place of its parent in the tree.

sipa commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,597 @@
+spend a Taproot single-sig output about 1% more expensive than P2WPKH.
+However, the costs are distributed differently---it costs more to create
+a Taproot output than a P2WPKH output and less to spend it, which may
+help contribute towards keeping the size of the UTXO set manageable.
```
It's worth comparing with native P2WSH where people are already fine with sending to a 32-byte witness program (and I don't think I've ever heard someone complain about that). Also, one idea for a change to the BIP (something @jnewbery brought up) is dropping the 0/1 prefix byte and just putting the X coordinate in the output. Obviously not something to put in the description right now, but it may help color things.

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
> I'd probably calculate Hash_TapLeaf("OP_RETURN OP_PUSH32 <256 bit random number>") to be sure it's unspendable.

I originally wrote that, but on reflection, it seemed to me that the probability of a digest of a random number being the preimage for a known spendable script is the same as the probability of a random number itself being the preimage for a known spendable script.  I'd love to know if I reasoned that incorrectly, though.

> Doing Hash_TapLeaf("<256bit rand>") could end up spendable

I was thinking of just using "<256bit rand>" in the branch construction so that it doesn't need to hashed (a useful optimization if you have to fill in a lot of blanks in your tree).  E.g., if 0x1234...cdef is the hash of a leaf for which I don't have a script to put in its pair sibling, I'd do `Hash_TapBranch(sort("0x1234...cdef", random_bytes(32)))`.

OTOH, having thought about it a bit more while writing this, I guess you could just fill in blank spots like that by duplicating a functional leaf you already had.  That would cut down on the storage requirements and wouldn't increase the amount of hashing necessary.  Though I guess it might be less private that way if you paired a leaf with itself.

sipa replied:
The advantage of creating unspendable leaves as an actual unspendable script rather than a random value is that you can prove to other participants (co-signers, ...) that it is in fact an unspendable branch.

harding replied:
Ah!  Thanks!

harding commented on _posts/en/newsletters/2019-05-14-newsletter.md:
I'm happy to remove it, but I was thinking of the LN funding transaction: https://github.com/lightningnetwork/lightning-rfc/blob/master/03-transactions.md#funding-transaction-output

jnewbery replied:
I'd prefer to remove it. LN has lots of different transaction output scriptPubKeys, so unless you specify this as a 'LN funding transaction', I don't think it makes things any clearer.

practicalswift commented on _posts/en/newsletters/2019-05-14-newsletter.md:
The `<B` is parsed as HTML (bold) which messes up the rendered version. Perhaps `&lt;B` can be used instead.

harding replied:
@practicalswift I suggest you look at the version rendered by Netlify rather than GitHub, since all the links and images and other site-specific stuff will work there.  (In the GitHub status box near the bottom of this page, expand the "All checks has passed" item and click the *Details* link to the right of "Deploy preview is ready".)

harding commented:
The newsletter is now complete and I believe I've addressed all feedback.  However, I'm still happy to make more edits if anyone is crazy enough to read it again.

My deepest thanks to each of you for all of your excellent review comments and for your work, both past and present, on this amazing improvement proposal.

practicalswift commented on _posts/en/newsletters/2019-05-14-newsletter.md:
```diff
@@ -0,0 +1,648 @@
+| Private key      | (Same as above)                         | `0x807d[...]0101` |
+| Public key       | (Same as above)                         | `0x02e5[...]3c23` |
+| Alter key prefix | (key[0] - 2),key[1:33])                 | `0x00e5[...]3c23` |
+| Address          | bech32.encode('bc', 1, 0x00e5[...]3c23) | `bc1pqr[...]xg73` |
```
What about including the full private key `0x807d[...]0101` to allow for using all the examples in the post as test vectors? I think that would be helpful.

harding replied:
@practicalswift test vectors are good, but I don't want to encourage people to write implementations based on an overview of the proposals.  If anyone wants to write a toy implementation now, I'd highly encourage them to try to read only the proposed BIPs; that way they can identify any confusing sections or omitted details or other problems in what will be the primary sources for Taproot and Tapscript details---and they can then ask the BIP author to make edits while the document is still a discussion draft.

jnewbery commented on _posts/en/newsletters/2019-05-14-newsletter.md:
@harding - I'd suggest dropping "(e.g. bech32 address)" unless you think it makes this sentence a lot easier to understand.

jnewbery commented on img/posts/segwit-savings.gnuplot:
```diff
@@ -3,6 +3,7 @@
 set style line 1 lc rgb '#8b1a0e' pt 4 ps 0.25 lt 1 lw 2
 set style line 2 lc rgb '#5e9c36' pt 4 ps 0.25 lt 1 lw 2
 set style line 3 lc rgb '#0025ad' pt 4 ps 0.25 lt 1 lw 2
+set style line 4 lc rgb '#7e2f8e' pt 4 ps 0.25 lt 1 lw 2
```
blue-purple isn't the clearest contrast (eg for people with certain color blindness). Could you be persuaded to change this to purple-yellow or something else with more contrast?

jnewbery commented on _posts/en/newsletters/2019-05-14-newsletter.md:
I don't like that we have "fee density (fee/vbyte)" in the previous sentence and "fee density (feerate)" here. Suggested change: "fee density (fee/vbyte) and sigop density (sigops/vbyte).  This is much more difficult than optimizing block composition based on just fee density."

jnewbery commented:
A book's worth of content with the humility of being a newsletter.

ACK a763cea49df685b2cc5296ae7176a7a128ce4fac.
